File: examples/chatapp/node-be/src/worker.ts
import * as db from "./db.ts";
import { type AppCtx, FANOUT_QUEUE, FAIL_QUEUE, REAP_QUEUE, schedulerMs } from "./context.ts";

interface MessageJob {
  message_id: string;
}

type Stopped = () => boolean;

const VISIBILITY = 30;
const WAIT = 1;

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

function signalFromStopped(stopped: Stopped): AbortSignal {
  const stop = new AbortController();
  const timer = setInterval(() => {
    if (stopped()) {
      stop.abort();
      clearInterval(timer);
    }
  }, 500);
  timer.unref();
  return stop.signal;
}

// fanout: marks each recipient's receipt delivered, idempotent on message id.
// Unread derives from receipts.read_at, so there is no counter to bump.
async function handleFanout(app: AppCtx, payload: MessageJob): Promise<void> {
  const { message_id } = payload;
  const msg = await db.messageById(app.pool, message_id);
  if (!msg) return; // disappeared before delivery
  const recipients = await db.otherMemberIds(app.pool, msg.chat_id, msg.sender_id);
  for (const uid of recipients) {
    await db.markDelivered(app.pool, message_id, uid);
  }
}

export function runFanoutWorker(app: AppCtx, stopped: Stopped): void {
  void app.forge.worker<MessageJob>(
    FANOUT_QUEUE,
    async (job) => handleFanout(app, job.payload),
    { visibilitySeconds: VISIBILITY, waitSeconds: WAIT, signal: signalFromStopped(stopped) },
  );
}

// reap: hard-deletes a disappearing message's row + blob when its scheduled job
// fires. Blob goes BEFORE the row, and a blob-delete failure propagates so the
// at-least-once queue redelivers instead of orphaning the object. Idempotent:
// already-gone or recalled (toggled off / not yet due) messages succeed cleanly.
async function reapMessage(app: AppCtx, payload: MessageJob): Promise<void> {
  const { message_id } = payload;
  const target = await db.reapTarget(app.pool, message_id);
  if (!target) return; // already gone
  // expires_at cleared (toggled off) or in the future (recalled / not yet due).
  if (target.expires_at === null || target.expires_at.getTime() > Date.now()) return;
  if (target.media_key) {
    // Propagate on failure -> nack -> redelivery, rather than orphan the blob.
    await app.forge.blobDelete(target.media_key);
  }
  await db.deleteIfDue(app.pool, message_id);
}

export function runReapWorker(app: AppCtx, stopped: Stopped): void {
  void app.forge.worker<MessageJob>(
    REAP_QUEUE,
    async (job) => reapMessage(app, job.payload),
    { visibilitySeconds: VISIBILITY, waitSeconds: WAIT, signal: signalFromStopped(stopped) },
  );
}

// fail: always nacks, so a triggered job exhausts its single attempt and
// dead-letters into `fail.dlq` (the opsStats DLQ demo).
export function runFailWorker(app: AppCtx, stopped: Stopped): void {
  void app.forge.worker<string>(
    FAIL_QUEUE,
    async () => {
      throw new Error("intentional failure for DLQ demo");
    },
    {
      visibilitySeconds: VISIBILITY,
      waitSeconds: WAIT,
      signal: signalFromStopped(stopped),
    },
  );
}

// Reconciliation heals work dropped after commit. The app and Forge hold separate
// pools, so the post-commit enqueue/schedule can't share the message's tx; a crash
// in that window leaves a message with no reap or no fanout. Each tick sweeps a
// bounded batch to self-heal both, idempotently.
const RECONCILE_LIMIT = 100;

async function reconcileOnce(app: AppCtx): Promise<void> {
  // Dropped reaps: delete due disappearing messages and their blobs. Blob delete is
  // best-effort here (the scheduled reap is the durable path); the row delete is
  // idempotent on expires_at <= now().
  for (const m of await db.dueMessages(app.pool, RECONCILE_LIMIT)) {
    if (m.media_key) {
      try {
        await app.forge.blobDelete(m.media_key);
      } catch {
        /* best-effort; deleteIfDue still clears the row */
      }
    }
    await db.deleteIfDue(app.pool, m.id);
  }
  // Dropped fanout: re-enqueue fanout for older messages with undelivered receipts.
  // Fanout is idempotent on mark_delivered, so a re-enqueue is always safe.
  for (const id of await db.undeliveredMessageIds(app.pool, RECONCILE_LIMIT)) {
    await app.forge.queue<MessageJob>(FANOUT_QUEUE).enqueue({ message_id: id }, { dedupId: id });
  }
}

// scheduler + reconciliation tick: fire due `at` jobs into their queue, then heal
// any work the post-commit enqueue/schedule dropped.
export function runScheduler(app: AppCtx, stopped: Stopped): void {
  const tick = schedulerMs();
  void (async () => {
    while (!stopped()) {
      try {
        await app.forge.runSchedulerOnce();
        await reconcileOnce(app);
        // Sweep expired Forge storage rows.
        await app.forge.maintain();
      } catch (e) {
        console.warn("scheduler tick failed:", (e as Error).message);
      }
      await sleep(tick);
    }
  })();
}

File: examples/chatapp/rust-be/src/worker.rs
//! In-process queue workers. All run on Forge's at-least-once queue, so every handler
//! is idempotent on the message id. Safe to run on every replica.

use std::future::Future;
use std::time::Duration;

use forgelib::{Bytes, EnqueueOpts, Job};

use crate::context::{Ctx, FAIL_QUEUE, FANOUT_QUEUE, MessageJob, REAP_QUEUE};
use crate::db;

/// Marks each recipient's receipt delivered. Idempotent: `mark_delivered` flips
/// `delivered_at` at most once, so redelivery is a no-op. Unread is derived from
/// receipts at read time, so there is nothing else to do here.
pub async fn run_fanout(ctx: Ctx, shutdown: impl Future<Output = ()> + Send) {
    let c = ctx.clone();
    ctx.forge
        .worker(FANOUT_QUEUE)
        .concurrency(4)
        .poll_wait(Duration::from_millis(200))
        .run_until(shutdown, move |job: Job| {
            let c = c.clone();
            async move { fanout(c, job).await }
        })
        .await;
}

/// Always errors, so a single-attempt job dead-letters into `fail.dlq`.
pub async fn run_fail(ctx: Ctx, shutdown: impl Future<Output = ()> + Send) {
    ctx.forge
        .worker(FAIL_QUEUE)
        .poll_wait(Duration::from_millis(200))
        .run_until(shutdown, move |_job: Job| async move {
            Err(anyhow::anyhow!("intentional failure (DLQ demo)"))
        })
        .await;
}

/// Reaps a disappearing message when `forge schedule` fires its expiry: delete the
/// row and its blob. Idempotent on an already-gone message.
pub async fn run_reap(ctx: Ctx, shutdown: impl Future<Output = ()> + Send) {
    let c = ctx.clone();
    ctx.forge
        .worker(REAP_QUEUE)
        .poll_wait(Duration::from_millis(200))
        .run_until(shutdown, move |job: Job| {
            let c = c.clone();
            async move { reap(c, job).await }
        })
        .await;
}

async fn fanout(c: Ctx, job: Job) -> anyhow::Result<()> {
    let fan: MessageJob = job
        .payload_json()
        .map_err(|e| anyhow::anyhow!("bad fanout payload: {e}"))?;

    let Some(msg) = db::message(&c.pool, fan.message_id).await? else {
        // Message vanished (e.g. disappeared) before delivery; not an error.
        return Ok(());
    };

    for uid in db::other_member_ids(&c.pool, msg.chat_id, msg.sender_id).await? {
        db::mark_delivered(&c.pool, fan.message_id, uid).await?;
    }
    Ok(())
}

async fn reap(c: Ctx, job: Job) -> anyhow::Result<()> {
    let r: MessageJob = job
        .payload_json()
        .map_err(|e| anyhow::anyhow!("bad reap payload: {e}"))?;
    let Some((media_key, expires_at)) = db::message_reap_info(&c.pool, r.message_id).await? else {
        return Ok(()); // already gone
    };
    // Not due (toggled off / recalled, or not yet expired): leave it alone.
    if expires_at.is_none_or(|e| e > chrono::Utc::now()) {
        return Ok(());
    }
    // Delete the blob before the row, and propagate failure so the at-least-once queue
    // redelivers rather than orphaning the object.
    if let Some(key) = media_key {
        c.forge.blob().delete(&key).await?;
    }
    db::delete_expired_message(&c.pool, r.message_id).await?;
    Ok(())
}

/// Heal work whose post-commit enqueue/schedule was dropped. The app and Forge hold
/// separate pools, so a message commit and its queue/schedule call can't share a tx;
/// a crash between them strands the follow-up. Run once per scheduler tick, bounded.
pub async fn reconcile(c: &Ctx) -> anyhow::Result<()> {
    // Dropped reaps: a due disappearing message whose reap never fired. Delete blob
    // (best-effort here; the bounded sweep re-runs next tick) then the row.
    for (id, media_key) in db::due_messages(&c.pool, 100).await? {
        if let Some(key) = media_key {
            let _ = c.forge.blob().delete(&key).await;
        }
        db::delete_expired_message(&c.pool, id).await?;
    }
    // Dropped fanout: a message past the grace window with never-delivered receipts.
    // Re-enqueue with a dedup id so a still-pending original can't double up; fanout is
    // idempotent on `mark_delivered` regardless.
    for id in db::undelivered_message_ids(&c.pool, 100).await? {
        let payload = serde_json::to_vec(&MessageJob { message_id: id })?;
        c.forge
            .queue()
            .enqueue(
                FANOUT_QUEUE,
                Bytes::from(payload),
                EnqueueOpts::new().with_dedup_id(id.to_string()),
            )
            .await?;
    }
    Ok(())
}
